pandas/core/sparse/array.py

```diff
@@ -707,41 +1064,204 @@ def mean(self, axis=0, *args, **kwargs):
-            Don't include counts of NaN, even if NaN is in sp_values.
+    @property
+    def T(self):
+        """Returns the SparseArray."""
+        return self
+
+    # ------------------------------------------------------------------------
+    # Ufuncs
+    # ------------------------------------------------------------------------
+
+    def __array_wrap__(self, array, context=None):
+        from pandas.core.dtypes.generic import ABCSparseSeries
+
+        ufunc, inputs, _ = context
+        inputs = tuple(x.values if isinstance(x, ABCSparseSeries) else x
+                       for x in inputs)
+        return self.__array_ufunc__(ufunc, '__call__', *inputs)
+
+    _HANDLED_TYPES = (np.ndarray, numbers.Number)
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        out = kwargs.get('out', ())
+
+        for x in inputs + out:
+            if not isinstance(x, self._HANDLED_TYPES + (SparseArray,)):
+                return NotImplemented
+
+        special = {'add', 'sub', 'mul', 'pow', 'mod', 'floordiv', 'truediv',
+                   'divmod', 'eq', 'ne', 'lt', 'gt', 'le', 'ge', 'remainder'}
+        if compat.PY2:
+            special.add('div')
+        aliases = {
+            'subtract': 'sub',
+            'multiply': 'mul',
+            'floor_divide': 'floordiv',
+            'true_divide': 'truediv',
+            'power': 'pow',
+            'remainder': 'mod',
+            'divide': 'div',
+        }
+        op_name = ufunc.__name__
+        op_name = aliases.get(op_name, op_name)
+
+        if op_name in special and kwargs.get('out') is None:
+            if isinstance(inputs[0], type(self)):
+                return getattr(self, '__{}__'.format(op_name))(inputs[1])
+            else:
+                return getattr(self, '__r{}__'.format(op_name))(inputs[0])
+
+        if len(inputs) == 1:
+            # No alignment necessary.
+            sp_values = getattr(ufunc, method)(self.sp_values, **kwargs)
+            fill_value = getattr(ufunc, method)(self.fill_value, **kwargs)
+            return self._simple_new(sp_values,
+                                    self.sp_index,
+                                    SparseDtype(sp_values.dtype, fill_value))
+
+        result = getattr(ufunc, method)(*[np.asarray(x) for x in inputs],
+                                        **kwargs)
+        if out:
+            if len(out) == 1:
+                out = out[0]
+            return out
+
+        if type(result) is tuple:
+            return tuple(type(self)(x) for x in result)
+        elif method == 'at':
+            # no return value
+            return None
+        else:
+            return type(self)(result)
+
+    def __abs__(self):
+        return np.abs(self)
+
+    # ------------------------------------------------------------------------
+    # Ops
+    # ------------------------------------------------------------------------
+
+    @classmethod
+    def _create_unary_method(cls, op):
```

Comment:
for reviewing purposes, is the code below to create those methods new code? Or copied from somewhere else?

Reply:
I believe new code. Previously, this would have been handled by `ndarray` and funds (I think).